test/org/javarosa/xpath/expr/Scenario.java

```diff
@@ -0,0 +1,490 @@
+     * because it would return an element corresponding to the xPath "/data"
+     * ("data" is commonly used as the main instance's xml tag), not the root
+     * element.
+     */
+    private TreeElement resolve(String xPath) {
+        // Get the real root element
+        TreeElement root = getRootElement();
+        // Since we start searching from "/", we make the input xPath relative to that
+        String relativeXPath = xPath.startsWith("/") ? xPath.substring(1) : xPath;
+
+        return resolve(root, relativeXPath);
+    }
+
+    private TreeElement getRootElement() {
+        return (TreeElement) formDef.getMainInstance().getRoot().getParent();
+    }
+
+    /**
+     * Returns the element corresponding to the given xPath.
+     * <p>
+     * It does so by recursively traversing children of the given element and calling
+     * {@link TreeElement#getChild(String, int)} on them.
+     */
+    private TreeElement resolve(TreeElement element, String xPath) {
+        List<String> parts = Arrays.asList(xPath.split("/"));
+        String firstPart = parts.get(0);
+        TreeElement nextElement = element.getChild(parseName(firstPart), parseMultiplicity(firstPart));
+
+        // Return null when a child with the given name and multiplicity doesn't exist.
+        if (nextElement == null)
+            return null;
+
+        // If there are more parts to analyze, call recursively on child
+        if (parts.size() > 1)
+            return resolve(nextElement, shift(xPath));
+
+        // If this is the last part in the xPath, we have the element we're looking for
+        return nextElement;
+    }
+
+    private static String shift(String xPath) {
+        List<String> parts = Arrays.asList(xPath.split("/"));
+        return String.join("/", parts.subList(1, parts.size()));
+    }
+
+    private static String pop(String xPath) {
+        List<String> parts = Arrays.asList(xPath.split("/"));
+        return String.join("/", parts.subList(0, parts.size() - 1));
+    }
+
+    private static String tailPart(String xPath) {
+        List<String> parts = Arrays.asList(xPath.split("/"));
+        return parts.get(parts.size() - 1);
+    }
+
+    private Optional<TreeElement> getFirstDescendantWithName(TreeElement node, String name) {
+        if (isNotRoot(node) && isNotTemplate(node) && elementMatchesName(node, name))
+            return Optional.of(node);
+
+        List<TreeElement> nonTemplateChildren = childrenOf(node).stream()
+            .filter(this::isNotTemplate)
+            .collect(Collectors.toList());
+
+        for (TreeElement child : nonTemplateChildren) {
+            Optional<TreeElement> result = getFirstDescendantWithName(child, name);
+            if (result.isPresent())
+                return result;
+        }
+
+        return Optional.empty();
+    }
+
+    private List<TreeElement> childrenOf(TreeElement node) {
+        List<TreeElement> children = new ArrayList<>();
```

Comment:
It probably doesn’t matter for these tests, but it might set a better example if you gave an `initialCapacity`, since you know how many elements there will be.